bip-musig2.mediawiki

```diff
@@ -313,17 +314,26 @@ Input:
 ==== Nonce Generation ====
 
 Input:
-* The secret signing key ''sk'': a 32-byte array or 0-byte array (optional argument)
-* The aggregate public key ''aggpk'': a 32-byte array or 0-byte array (optional argument)
-* The message ''m'': a 32-byte array or 0-byte array (optional argument)
+* The secret signing key ''sk'': a 32-byte array (optional argument)
+* The aggregate public key ''aggpk'': a 32-byte array (optional argument)
+* The message ''m'': a byte array with ''0 &le; len(m) &le; 2<sup>32</sup>-1'' (optional argument)
 * The auxiliary input ''extra_in'': a byte array with ''0 &le; len(extra_in) &le; 2<sup>32</sup>-1'' (optional argument)
 
 '''''NonceGen(sk, aggpk, m, extra_in)''''':
 * Let ''rand' '' be a 32-byte array freshly drawn uniformly at random
-* If ''len(sk) > 0'':
+* If the optional argument ''sk'' is present:
 ** Let ''rand'' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand')''<ref>The random data is hashed (with a unique tag) as a precaution against situations where the randomness may be correlated with the secret signing key itself. It is xored with the secret key (rather than combined with it in a hash) to reduce the number of operations exposed to the actual secret key.</ref>.
 * Else: let ''rand = rand' ''
-* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(1, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
+* If the optional argument ''aggpk'' is not present:
+** Let ''aggpk = empty_bytestring''
+* If the optional argument ''m'' is not present:
+** Let ''m = empty_bytestring''
+** Let ''m_present = bytes(1, 0)''
+* Else:
+** Let ''m_present = bytes(1, 1)''
+* If the optional argument ''extra_in'' is not present:
+** Let ''extra_in = empty_bytestring''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || m_present || bytes(8, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i - 1))) mod n'' for ''i = 1,2''
```

Comment:
Nit: Would it be worth adding a footnote explaining why `0 <= len(m) <= 2^32 - 1` but we use 8 bytes for encoding `len(m)`?

Reply:
Good point, this doesn't make sense. I think it's best if we take the same approach with respect to message length as https://github.com/sipa/bips/pull/221.

Reply:
Actually in contrast to BIP-340 we need to restrict the message length because we have to serialize it within 8 bytes.

Reply:
I added a commit that extends the message size and adds a footnote that explains that the SHA256 precondition implicitly propagates to the caller. That looks quite weird. It seems to me that it'd be easier here if we'd restric the size of messages in bip340 to 2^48-1 or something.